sdk/storage/azure-storage-blob/src/samples/java/com/azure/storage/blob/BufferedUploadWithKnownLengthExample.java

```diff
@@ -0,0 +1,103 @@
+     * @throws RuntimeException If the downloaded data doesn't match the uploaded data
+     */
+    public static void main(String[] args) throws IOException {
+
+        /*
+         * For more information on this setup, please refer to the BasicExample.
+         */
+        String accountName = SampleHelper.getAccountName();
+        String accountKey = SampleHelper.getAccountKey();
+        StorageSharedKeyCredential credential = new StorageSharedKeyCredential(accountName, accountKey);
+        String endpoint = String.format(Locale.ROOT, "https://%s.blob.core.windows.net", accountName);
+        String containerName = "myjavacontainerbuffereduploadlength" + System.currentTimeMillis();
+        BlobServiceAsyncClient storageClient = new BlobServiceClientBuilder().endpoint(endpoint).credential(credential)
+            .buildAsyncClient();
+
+        BlobContainerAsyncClient containerClient = storageClient.getBlobContainerAsyncClient(containerName);
+        containerClient.create().block();
+
+        uploadSourceBlob(endpoint, credential, containerName);
+        BlobAsyncClient blobClient = containerClient.getBlobAsyncClient("HelloWorld.txt");
+
+
+        /*
+        sourceData has a network stream as its source and therefore likely does not support multiple subscribers. Even
+        if it did support multiple subscribers, it would not produce the same data each time it was subscribed to. While
+        we could inspect the http headers for the content-length, let us suppose that this information is unavailable
+        at this time. All three of these factors would individually make the use of the standard upload method
+        impossible--the first two because retries would not work and the third one because we could not satisfy the
+        argument list.
+         */
+        Flux<ByteBuffer> sourceData = getSourceBlobClient(endpoint, credential, containerName).downloadStream()
+            // Perform transformation with length of 1 GB.
+            .map(BufferedUploadWithKnownLengthExample::bufferTransformation);
+
+        /*
+        Although this upload overload permits the use of such unreliable data sources, with known length we can speed
+        up the upload process. A buffer size and maximum concurrency can still be passed in to achieve optimized upload.
+         */
+        long length = 10;
+        long blockSize = 10 * 1024;
+        int maxConcurrency = 5;
+        ParallelTransferOptions parallelTransferOptions = new ParallelTransferOptions()
+            .setBlockSizeLong(blockSize)
+            .setMaxConcurrency(maxConcurrency);
+
+        // Since we already know the size of our buffered bytes, we can pass the ByteBuffer and length to the BinaryData.
+        // This will internally convert the BinaryData to a Flux<ByteBuffer>, but with known length we can optimize the
+        // upload speed.
+        BinaryData binaryData = BinaryData.fromFlux(sourceData, length).block();
```

Comment:
Two minor nits:

- This returns a reactive response and can be reactively chained
- There is two Flux<ByteBuffer> length APIs in BinaryData, the one being used here will buffer the stream whereas the overload will allow this to be configured. We should call that out here